src/braket/default_simulator/linalg_utils.py

```diff
@@ -0,0 +1,109 @@
+    axes = (
+        np.arange(len(targets), 2 * len(targets)),
+        targets,
+    )
+    product = np.tensordot(gate_matrix, state, axes=axes)
+
+    # Axes given in `operation.targets` are in the first positions.
+    unused_idxs = [idx for idx in range(len(state.shape)) if idx not in targets]
+    permutation = list(targets) + unused_idxs
+    # Invert the permutation to put the indices in the correct place
+    inverse_permutation = np.argsort(permutation)
+    return np.transpose(product, inverse_permutation)
+
+
+def marginal_probability(
+    probabilities: np.ndarray,
+    targets: List[int] = None,
+) -> np.ndarray:
+    """Return the marginal probability of the computational basis states.
+
+    The marginal probability is obtained by summing the probabilities on
+    the unused qubits. If no targets are specified, then the probability
+    of all basis states is returned.
+
+    Args:
+        probabilities (np.ndarray): The probability distribution to marginalize.
+        targets (List[int]): The qubits of the marginal distribution;
+            if no targets are specified, then the probability of all basis states is returned.
+
+    Returns:
+        np.ndarray: The marginal probability distribution.
+    """
+    qubit_count = int(np.log2(len(probabilities)))
+
+    if targets is None or targets == list(range(qubit_count)):
+        # All qubits targeted, no need to marginalize
+        return probabilities
+
+    targets = np.hstack(targets)
+
+    # Find unused qubits and sum over them
+    unused_qubits = list(set(range(qubit_count)) - set(targets))
+    as_tensor = probabilities.reshape([2] * qubit_count)
+    marginal = np.apply_over_axes(np.sum, as_tensor, unused_qubits).flatten()
+
+    # Reorder qubits to match targets
+    basis_states = np.array(list(itertools.product([0, 1], repeat=len(targets))))
+    perm = np.ravel_multi_index(
+        basis_states[:, np.argsort(np.argsort(targets))].T, [2] * len(targets)
+    )
+    return marginal[perm]
+
+
+def partial_trace(
+    density_matrix: np.ndarray,
+    qubit_count: int,
+    targets: Optional[List[int]] = None,
+) -> np.ndarray:
+    """Returns the reduced density matrix for the target qubits.
+
+    If no target qubits are supplied, this method returns the trace of the density matrix.
+
+    Args:
+        density_matrix (np.ndarray): The density matrix to reduce.
+        qubit_count (int): The number of qubits in the density matrix.
+        targets (List[int]): The qubits of the output reduced density matrix;
+            if no target qubits are supplied, this method returns the trace of the density matrix.
+
+    Returns:
+        np.ndarray: The partial trace of the density matrix.
+    """
+    target_set = set(targets) if targets else set()
+    nkeep = 2 ** len(target_set)
+    idx1 = [i for i in range(qubit_count)]
+    idx2 = [qubit_count + i if i in target_set else i for i in list(range(qubit_count))]
```

Comment:
```suggestion
    idx2 = [qubit_count + i if i in target_set else i for i in range(qubit_count)]
```

Reply:
Why the extra list instantiation? The range object achieves the goal

Reply:
Exactly, hence my suggestion :)

Reply:
Oh right, suggestion comes after